### PR TITLE
Add logging to autodiff

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -837,13 +867,21 @@ static void lambdaLiftReverse(Gradient& grad_desc, ReverseDetails& rev_info) {
...
+        new_input->debugName(),
+        " for an embedded backprop block");
   };
   for (auto& offset : grad_desc.df_input_captured_inputs)
     add_capture(graph.inputs()[offset]);
   for (auto& offset : grad_desc.df_input_captured_outputs)
     add_capture(graph.outputs()[offset]);
 
+  GRAPH_DUMP(" forward propagation: ", &graph);
+  GRAPH_DEBUG(" backward propagation: ", *(reverse_block->owningNode()));
```
Comment (apaszke) [inline torch/csrc/jit/autodiff.cpp]: nit: no one calls this forward propagation. Let's do either forward/backward or primal/reverse

**Review by apaszke — COMMENTED** [review]
Isn't the debug mode going to get overly verbose if there's no way to filter it by the component?

**Comment by Krovatkin** [comment]
@apaszke it does get pretty verbose, considering that `autodiff.cpp` also runs DCE and that in turn needs alias analysis. @zdevito doesn't want any extra options controlling, so I'm putting together a very simple script that generates an html page with collapsible sections from a log file, I'm planning on keeping things like alias analysis and DCE collapsed by default.

File: torch/csrc/jit/autodiff.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/autodiff.cpp]: thanks! will fix!

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -676,7 +690,13 @@ static void deduplicateSizeCaptures(
     }
     if (usedOnlyInReverse(capture) && capture_set.count(node->input())) {
       WithInsertPoint insert_guard{*rev_info.reverse_block->nodes().begin()};
-      capture->replaceAllUsesWith(SymbolicVariable(node->input()).size());
+      auto size = SymbolicVariable(node->input()).size();
```
Comment (ailzhang) [inline torch/csrc/jit/autodiff.cpp]: Is `SymbolicVariable` necessary here?

**Review by ailzhang — APPROVED** [review]
As our logging gets more verbose, maybe it's good to have the debug info categorized like VLOG(n)? 
For example if there's an easy way to turn on/off/filter out autodiff debug messages, it might be helpful. 
But it could be a followup diff if you think it's reasonable. :D

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/autodiff.cpp]: I don't think so; Mike should be removing these shortly!

**Comment by apaszke** [comment]
Wow, an HTML page would be amazing! However, if we want to go down the path of "this is not too much for a human to process without help" anyway, why won't we make our logging API more structured (i.e. encode the HTML structure in the API instead of parsing the output later)?

**Comment by Krovatkin** [comment]
> if we want to go down the path of "this is not too much for a human to process without help" anyway
it's already probably a little bit too verbose
> why won't we make our logging API more structured 
we could consider making it more structured. However, adding more structure will make the logging output more difficult to parse and we will pretty much limit everyone to only **one** way of reading logs. HTML is both notoriously hard to parse (even if it's well-formed) and read ("<", ">", "'", etc need to be escaped). 

@ZolotukhinM , I think you already tried using logging? What are you thoughts on this?

FYI: @zdevito @suo

**Comment by ZolotukhinM** [comment]
I think using HTML for output would only make sense if we have knobs to control what we exactly  output (i.e. what module/function, what pass, etc.). Outputting HTML always without an option to turn it off would not be very convenient as it's harder to filter for instance.

From my experience the following use-cases should be covered:
1) I'd like to see how IR of a given function evolves through the compilation process.
2) I'd like to find why my pass is doing something funky (i.e. debug).
3) I'd like to collect some stats about the compilation process (e.g. check if a transformation was triggered or not, or how many times it fired).

In all these cases I find it extremely useful if I can see only logging for a function I'm interested in and from the passes I'm interested in. I imagine that can be achieved either by extra flags or by making the outputs simple to parse and filter afterwards. If we decide to add flags, then adding a flag for specifying output format would also make sense.

FWIW, in LLVM it was solved by extensive configurability through flags, but in LLVM there was a standalone pass runner (opt) that naturally allowed that.
